### PR TITLE
Change test selection algorithm to catch edge case

### PR DESCRIPTION
**Pull request by Genki-S**
## Background

Current test selection algorithm is like this: "When a code is changed, and the previous code is executed by certain test cases, then those test cases are selected."

## Problem

The problem with this algorithm is that there are cases it cannot select affected test cases. For example, when this code:

```
def double(n)
  # Lorem
  # ipsum
  # dolor
  n * 2
end
```

is changed to this:

```
def double(n)
  # Lorem
  n *= 2
  # dolor
  n * 2
end
```

TTNT cannot select test which test this method since current algorithm selects line 3 of original code as target line, and the line 3 of original code (line of `# ipsum`) is not executed by the test.

## Solution

So, I have changed test selection algorithm like this:

"When a line in a file is changed, for a given test, if the changed line is included in the range of coverage (i.e. the topmost executed line and the bottommost executed line), then the test is selected."

I came up with some heuristics (like selecting tests from nearest line which has coverage information), but finally settled on this since I want to value "reliability" on TTNT. That is, I think it's OK to select redundant tests, but I think it's not OK to fail to select tests which become failing after some changes. This heuristic I implemented, in my mind, is not likely to fail to select affected tests (though it will choose unaffected tests more).

## Changes

I changed test structure a bit in order to make it possible to add new fixture (first 4 commits): https://github.com/Genki-S/ttnt/commit/18b2f436cf454c1381c81af12a1eb5ee1dc9260d https://github.com/Genki-S/ttnt/commit/0a426a7e7f2350fb0b4cb8ced31f2a5b12889624 https://github.com/Genki-S/ttnt/commit/bdfb62cc23b22bdc2e7f0a7d3300b69be86feedb https://github.com/Genki-S/ttnt/commit/262497d9d017dedea039e9ddd90e5bdb708bd0fe

And I wrote a test case which fails with current test selection algorithm (with new fixtures): https://github.com/Genki-S/ttnt/commit/0e8e01de788bac4ce0b79f4fef39d193b1fdfdbd

Then I implemented new test selection algorithm: https://github.com/Genki-S/ttnt/commit/603652d7dd19dd1caf48b672b97470c1a3e07d7e

Finally, I fixed existing test cases to match the new test selection algorithm: https://github.com/Genki-S/ttnt/commit/6a6033600c36de35d49e4d7dfede7c52a5831df6

@robin850 could you review the idea and the code, please?

### CODE REVIEW

File: test/fixtures/fizzbuzz/fizzbuzz.rb [hunk 1]
```diff
@@ -1,10 +1,11 @@
+# Padding
 def fizzbuzz_convert(n)
   if n % 15 == 0
     "fizzbuzz"
-  elsif n % 3 == 0
-    "fizz"
   elsif n % 5 == 0
     "buzz"
+  elsif n % 3 == 0
+    "fizz"
```
Comment (Genki-S) [inline test/fixtures/fizzbuzz/fizzbuzz.rb]: I changed the order of "fizz" and "buzz" since there are test cases I want to test that TTNT just selectes test related to "fizz". If "fizz" line is above the "buzz" line, it means "fizz" line is "covered" by the coverage of "buzz_test.rb". Thus any change to "fizz" line leads to select "buzz_test.rb"...

File: test/fixtures/fizzbuzz/fizzbuzz.rb [hunk 2]
```diff
@@ -1,10 +1,11 @@
+# Padding
```
Comment (Genki-S) [inline test/fixtures/fizzbuzz/fizzbuzz.rb]: And I need to add newlines above this comment (and I need this comment) since adding newlines to the line of "fizzbuzz" leads to selecting both "fizz_test.rb" and "buzz_test.rb" with the same reason above.

**Comment by robin850** [comment]
Hi Genki !

This is looking good ; the only weak point here is that unrelated tests may be selected during the process but I agree with you that we must *at least* select tests that are related to the given change. However, this is a rather important concern for big files, mainly in Rails, it's common to see diffs like:

~~~ruby
require 'new/stuff'
  # existing methods and classes ...

  def method_that_relies_on_new_stuff_and_internal_method
  end

  private
    # exiting private methods ...

    def internal_method
    end
end
~~~

So with the current algorithm and such diff we are mostly selecting the tests for all the methods inside this file.

Let me think a bit more about that! Keep up the good work so far! :+1:

**Comment by Genki-S** [comment]
Hi Robin!

> the only weak point here is that unrelated tests may be selected during the process

> So with the current algorithm and such diff we are mostly selecting the tests for all the methods inside this file

Yes, this is my concern, too :disappointed: 
I want some clever algorithm to narrow the selection, but keep selecting necessary tests...

> Let me think a bit more about that!

That's very kind of you, thank you so much! :wink: I will think about it more, too!

**Comment by robin850** [comment]
Okay! After thinking a bit more about that, it looks like this is our best and simplest option for now! Even though we may select "unrelated" tests, that's still better to test the whole behavior anyway!

You can merge this then ; good work, as usual! :-)

**Comment by Genki-S** [comment]
OK, thank you! I hope we can improve on this later, if possible.
